substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java

```diff
@@ -186,33 +184,36 @@ public static InputStream createInputStream(String moduleName, String resourceNa
         if (resourceName == null) {
             return null;
         }
-        String canonicalResourceName = toCanonicalForm(resourceName);
 
         List<URL> resourcesURLs = new ArrayList<>();
 
-        /* If moduleName was unspecified we have to consider all modules in the image */
+        ResourceStorageEntry explicitEntry = Resources.get(moduleName, resourceName);
+        addURLEntries(resourcesURLs, explicitEntry, moduleName, resourceName);
+        /*
+         * If moduleName was unspecified we have to consider all modules in the image. We append
+         * those after the explicit entry in order to prioritize the explicit entry in case it
+         * exists
```

Comment:
@olpaw FYI

This change is a bit "out of scope" so I thought an explicit mention would help.

What it essentially does is to prioritize `resource:META-INF` over `resource:mymodule/META-INF` in case both are available and one requests `META-INF` without specifying a module.

Reply:
Thanks. I will have a look next week.

Reply:
> What it essentially does is to prioritize resource:META-INF over resource:mymodule/META-INF in case both are available and one requests META-INF without specifying a module.

Have you tested if this is the same prioritization that would happen when running on the jvm?

Reply:
Good question! No I did not.

Reply:
According to my tests JVM follows the old prioritization.

So I guess we should revert this part.

Wold you like me to update the PR @olpaw ?

Reply:
Yes, please.

Reply:
@zakkak ideally you would also add a test that checks that the order is identical to the order on jvm.

Reply:
> Yes, please.

Done

> @zakkak ideally you would also add a test that checks that the order is identical to the order on jvm.

I'll give that a try. However, since implementing the tests might take some time and what they test is unrelated to this PR I would like you to proceed with reviewing/merging this PR while I work on them as https://github.com/oracle/graal/issues/4326 is causing our pipelines to fail.

Are there any tests doing similar checks, i.e. comparing the jvm behavior to that of svm (so that I don't reinvent the wheel)?

Thanks

Reply:
> Are there any tests doing similar checks, i.e. comparing the jvm behavior to that of svm (so that I don't reinvent the wheel)?

Yes the whole `NativeImageResourceFileSystemProviderTest` tests are like that. Either run them on the JVM with
```
mx unittest --very-verbose com.oracle.svm.test.NativeImageResourceFileSystemProviderTest
```
or run them as native-image via
```
mx native-unittest ...
```
that's how I verify that our implementations behave exactly like the jvm.

Reply:
> PR I would like you to proceed with reviewing/merging this PR while I work on them as https://github.com/oracle/graal/issues/4326 is causing our pipelines to fail.

OK.